sql/inspect: add protected timestamp support to INSPECT

spilchen opened the pull request:
This change adds protected timestamp record management to INSPECT jobs when using AS OF SYSTEM TIME clauses. Protected timestamps prevent garbage collection of historical data during the inspection process.

When an INSPECT job specifies an AS OF timestamp, the job now:
- Creates a protected timestamp record before beginning inspection
- Stores the record ID in the job details for tracking
- Automatically cleans up the record on job completion or failure

Note: when we add support for running INSPECT without a fixed timestamp (in #148765), PTS won't be acquired then.

Closes #148573

Release note: none
Epic: CRDB-30356

rafiss reviewed (changes requested):
looks good overall! just want to make sure this works correctly if the Resume function returns early.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/153919#-:-Oa0YIcO2xbBO19SiFuG:bn2ksha)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @kev-cao)

___
*[`pkg/sql/inspect/inspect_job.go` line 74 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/153919#-Oa0XV6U-5Lo-YMmxOKx:-Oa0XV6U-5Lo-YMmxOKy:boj3qsu) ([raw file](https://github.com/cockroachdb/cockroach/blob/0dc358f067532902b6fb9c3c1a6319c74ad2f8d8/pkg/sql/inspect/inspect_job.go#L74)):*
> ```Go
> 	}
> 
> 	c.maybeCleanupProtectedTimestamp(ctx, execCfg)
> ```

if any of the above function calls return an error, does that mean we'd never reach this code to cleanup the PTS? i believe the desired pattern here would be to use `defer` immediately after the call to `maybeProtectTimestamp`.
___
*[`pkg/sql/inspect/inspect_job.go` line 254 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/153919#-Oa0X5IX1mnxubjcAhfx:-Oa0X5IY1jmL6GnTDPLV:bq1s1ux) ([raw file](https://github.com/cockroachdb/cockroach/blob/0dc358f067532902b6fb9c3c1a6319c74ad2f8d8/pkg/sql/inspect/inspect_job.go#L254)):*
> ```Go
> 
> 	// Protect will store the PTS ID in job details.
> 	_, err := execCfg.ProtectedTimestampManager.Protect(ctx, c.job, target, details.AsOf)
> ```

the call to `Protect` returns a `cleanup` function. is there any advantage to using that instead of the explicit call to `Unprotect`? perhaps the main benefit is that it means we don't need to access the field in the job details and check for its presence.
___
*[`pkg/sql/inspect/inspect_job.go` line 259 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/153919#-Oa0YB-ID-bzIk2-bLar:-Oa0YB-ID-bzIk2-bLas:b-7z5e1) ([raw file](https://github.com/cockroachdb/cockroach/blob/0dc358f067532902b6fb9c3c1a6319c74ad2f8d8/pkg/sql/inspect/inspect_job.go#L259)):*
> ```Go
> 	}
> 
> 	log.Dev.Infof(ctx, "protected timestamp created for INSPECT job %d", c.job.ID())
> ```

nit: as long as we have this debug log, it could be useful to include the timestamp itself here.


<!-- Sent from Reviewable.io -->

spilchen committed:
sql/inspect: add protected timestamp support to INSPECT

This change adds protected timestamp record management to INSPECT jobs
when using AS OF SYSTEM TIME clauses. Protected timestamps prevent
garbage collection of historical data during the inspection process.

When an INSPECT job specifies an AS OF timestamp, the job now:
- Creates a protected timestamp record before beginning inspection
- Stores the record ID in the job details for tracking
- Automatically cleans up the record on job completion or failure

Note: when we add support for running INSPECT without a fixed timestamp
(in #148675), PTS won't be acquired then.

Closes #148573

Release note: none
Epic: CRDB-30356

spilchen reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/153919)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @kev-cao and @rafiss)

___
*[`pkg/sql/inspect/inspect_job.go` line 74 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/153919#-Oa0XV6U-5Lo-YMmxOKx:-Oa0oyLu2VItWm1rNPm6:b-8w7vi9) ([raw file](https://github.com/cockroachdb/cockroach/blob/0dc358f067532902b6fb9c3c1a6319c74ad2f8d8/pkg/sql/inspect/inspect_job.go#L74)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

if any of the above function calls return an error, does that mean we'd never reach this code to cleanup the PTS? i believe the desired pattern here would be to use `defer` immediately after the call to `maybeProtectTimestamp`.
</blockquote></details>

I am relying on `OnFailOrCancel` to be called, which will unprotect the PTS if it's present in the job. I have a test for this too, which seems to confirm this. See the test in TestInspectJobProtectedTimestamp, which causes a failure in the call to `c.maybeRunAfterProtectedTimestampHook`.
___
*[`pkg/sql/inspect/inspect_job.go` line 254 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/153919#-Oa0X5IX1mnxubjcAhfx:-Oa0q42f3IgmIQiS8hj3:bf130mg) ([raw file](https://github.com/cockroachdb/cockroach/blob/0dc358f067532902b6fb9c3c1a6319c74ad2f8d8/pkg/sql/inspect/inspect_job.go#L254)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

the call to `Protect` returns a `cleanup` function. is there any advantage to using that instead of the explicit call to `Unprotect`? perhaps the main benefit is that it means we don't need to access the field in the job details and check for its presence.
</blockquote></details>

I wasn’t sure what the advantage was if we already have cleanup logic in `OnFailOrCancel`. In the other caller (`protectRestoreTargets` during resume), the cleanup function is ignored. By contrast, in `newSchemaChangeResumer`, `Unprotect` appears in `OnFailOrCancel`, but it’s labeled as a last resort. In that case, the main cleanup comes from the cleanup function in `Resume`.

So, I guess there are a couple of ways to handle this. I can flow the cleanup function up and call it at the end of `Resume` as an extra safeguard (like `newSchemaChanger`). Let me know if you prefer that approach.


<!-- Sent from Reviewable.io -->

rafiss reviewed (approved):
lgtm!

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/153919#-:-Oa0yJl83qQtly84c3uA:b-7nbn20)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @kev-cao and @spilchen)

___
*[`pkg/sql/inspect/inspect_job.go` line 74 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/153919#-Oa0XV6U-5Lo-YMmxOKx:-Oa0xxa97L-j32Gelzj_:b-e9c3lt) ([raw file](https://github.com/cockroachdb/cockroach/blob/0dc358f067532902b6fb9c3c1a6319c74ad2f8d8/pkg/sql/inspect/inspect_job.go#L74)):*
<details><summary><i>Previously, spilchen wrote…</i></summary><blockquote>

I am relying on `OnFailOrCancel` to be called, which will unprotect the PTS if it's present in the job. I have a test for this too, which seems to confirm this. See the test in TestInspectJobProtectedTimestamp, which causes a failure in the call to `c.maybeRunAfterProtectedTimestampHook`.
</blockquote></details>

ah, i missed the `OnFailOrCancel` changes. this seems safe then
___
*[`pkg/sql/inspect/inspect_job.go` line 254 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/153919#-Oa0X5IX1mnxubjcAhfx:-Oa0y2kL2zIg5vRalZvz:b1jhjkk) ([raw file](https://github.com/cockroachdb/cockroach/blob/0dc358f067532902b6fb9c3c1a6319c74ad2f8d8/pkg/sql/inspect/inspect_job.go#L254)):*
<details><summary><i>Previously, spilchen wrote…</i></summary><blockquote>

I wasn’t sure what the advantage was if we already have cleanup logic in `OnFailOrCancel`. In the other caller (`protectRestoreTargets` during resume), the cleanup function is ignored. By contrast, in `newSchemaChangeResumer`, `Unprotect` appears in `OnFailOrCancel`, but it’s labeled as a last resort. In that case, the main cleanup comes from the cleanup function in `Resume`.

So, I guess there are a couple of ways to handle this. I can flow the cleanup function up and call it at the end of `Resume` as an extra safeguard (like `newSchemaChanger`). Let me know if you prefer that approach.
</blockquote></details>

i don't see a need to change it. just was wondering if it made any of the code simpler, but it seems like not really.
___
*[`-- commits` line 14 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/153919#-Oa0xpRX4kT6hBugEUL6:-Oa0xpRX4kT6hBugEUL7:b-f37581):*
nit: this might be the wrong issue link
___
*[`pkg/sql/inspect/inspect_job_test.go` line 290 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/153919#-Oa0yGPH2OIiB7QH0JdA:-Oa0yGPH2OIiB7QH0JdB:b-lq9goe) ([raw file](https://github.com/cockroachdb/cockroach/blob/348fa55fdaa4c2bb8aa5a6ceb1b47f9ec627da08/pkg/sql/inspect/inspect_job_test.go#L290)):*
> ```Go
> 			require.Equal(t, tc.expectedJobStatus, status, "job should have expected status")
> 
> 			// Verify protected timestamp record is cleaned up
> ```

this test covers the concern i had, thanks!


<!-- Sent from Reviewable.io -->
